### PR TITLE
[SYCL] Added missing SYCL 2020 USM features

### PR DESCRIPTION
**Pull request by dnmokhov**
- USM atomic allocation aspects (+ tests)
- `handler::mem_advise`
- `handler::copy` and `queue::copy`
- `queue` class: overloads of `fill`, `memset`, `memcpy`, `copy`, `mem_advise`, `prefetch` with dependency event arguments
- optional property list argument in USM allocation functions and `usm_allocator` constructors (+ tests)

### CODE REVIEW

File: sycl/include/CL/sycl/queue.hpp [hunk 1]
```diff
@@ -320,17 +320,52 @@ class __SYCL_EXPORT queue {
...
+      CGH.depends_on(DepEvent);
+      CGH.fill<T>(Ptr, Pattern, Count);
+    });
+  }
+
+  /// Fills the specified memory with the specified pattern.
+  ///
+  /// \param Ptr is the pointer to the memory to fill.
+  /// \param Pattern is the pattern to fill into the memory.  T should be
+  /// trivially copyable.
```
Comment (cperkinsintel) [inline sycl/include/CL/sycl/queue.hpp:352]: If the Pattern is supposed to be trivially copyable, should this template be qualified with a `std::enable-if` clause?  

In Section 4.9.4.3 of the SYCL2020 spec it says that "T must be a scalar value or a SYCL vector type." for one of the other variants of `.fill()`     But for this variant (in 4.6.5.2) the spec has a one word link: "USM". 

@Pennycook do you know?

**Review by cperkinsintel — APPROVED** [review]
other than the question, LGTM.

File: sycl/include/CL/sycl/queue.hpp (hunk 1 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/queue.hpp:352]: The specification doesn't say "Only available if", which is SYCL 2020 speak for "Guarded by enable if".  So by the current specification, no.  @jbrodman, do you think there should be compile-time checking of `T` here?

Comment (dnmokhov) [reply sycl/include/CL/sycl/queue.hpp:352]: BTW, there is a `static_assert(is_trivially_copyable<T>)` in [the implementation of `handler::fill`](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/handler.hpp#L2151-L2152), which is called from here.

Comment (jbrodman) [reply sycl/include/CL/sycl/queue.hpp:352]: I think the static assert is fine.  I agree with John that enable_if isn't the right thing here.

**Comment by dnmokhov** [comment]
Pinging the reviewers: @intel/llvm-reviewers-runtime , @s-kanaev, @jbrodman, @dm-vodopyanov.

**Comment by vladimirlaz** [comment]
@dnmokhov could you add tests for newly added functionality to https://github.com/intel/llvm-test-suite/
If it is in place please add link to the PR description.

File: sycl/include/CL/sycl/aspects.hpp [hunk 2]
```diff
@@ -38,7 +38,9 @@ enum class aspect {
   ext_intel_gpu_subslices_per_slice = 22,
   ext_intel_gpu_eu_count_per_subslice = 23,
   ext_intel_max_mem_bandwidth = 24,
-  ext_intel_mem_channel = 25
+  ext_intel_mem_channel = 25,
+  usm_atomic_host_allocations = 26,
+  usm_atomic_shared_allocations = 27
```
Comment (s-kanaev) [inline sycl/include/CL/sycl/aspects.hpp:43]: NIT: suggest finishing this line with a comma.
```suggestion
  usm_atomic_shared_allocations = 27,
```

File: sycl/include/CL/sycl/queue.hpp [hunk 3]
```diff
@@ -339,6 +374,32 @@ class __SYCL_EXPORT queue {
...
+  ///
+  /// \param Ptr is a USM pointer to the memory to fill.
+  /// \param Value is a value to be set. Value is cast as an unsigned char.
+  /// \param Count is a number of bytes to fill.
+  /// \param DepEvent is an event that specifies the kernel dependencies.
+  /// \return an event representing fill operation.
+  event memset(void *Ptr, int Value, size_t Count, event DepEvent);
+
+  /// Fills the memory pointed by a USM pointer with the value specified.
+  /// No operations is done if \param Count is zero. An exception is thrown
```
Comment (cperkinsintel) [inline sycl/include/CL/sycl/queue.hpp:390]: This "no operation is done of Count is zero".  Is that check performed in the runtime? I didn't see that check.

Comment (dnmokhov) [reply sycl/include/CL/sycl/queue.hpp:390]: It is checked in [`MemoryManager::fill_usm`](https://github.com/intel/llvm/blob/82c915cc18717853921d9bfdcd9f95843230710a/sycl/source/detail/memory_manager.cpp#L664-L671), which is called by `queue_impl::memset`.

**Review by cperkinsintel — APPROVED** [review]
LGTM

File: sycl/include/CL/sycl/queue.hpp (hunk 3 above)
Comment (cperkinsintel) [reply sycl/include/CL/sycl/queue.hpp:390]: thanks - I cruised right past that.

**Review by s-kanaev — APPROVED** [review]
LGTM

**Comment by gmlueck** [comment]
I noticed that this change was just merged.

Coincidentally, the language evolution group just identified a spec bug here.  The USM `copy` member functions take the `src` and `dst` parameters in the wrong order.  The order should be `(src, dst)`, which matches the standard C++ `copy` function and matches the existing SYCL `copy` functions that take accessor parameters.

Can we change the implementation of these USM `copy` member functions?

**Comment by cperkinsintel** [comment]
@gmlueck  - is that right? `memcpy` is definitely dest pointer argument first, followed by source.   I thought `std::copy` took iterators, not pointers.   The `sycl::memcpy` and `sycl::copy` methods in this PR both take pointers, no?  Perhaps it's not the order of the arguments, but the naming of the `sycl::copy` function?

**Comment by gmlueck** [comment]
We agree that `std::memcpy` is `(dst, src)`, so we do not think the USM `memcpy` functions should change.  They should remain as they are with `(dst, src)` ordering.

The USM `copy` functions are the ones we're worried about, which also have the parameter order `(dst, src)`.  Our primary concern is that this does not match the parameter order for the `handler::copy()` member functions that take an `accessor` parameter.  Those functions have `(src, dst)` ordering.  We think it is particularly bad to have two overloads of the `handler::copy()` function, where one takes `(src, dst)` and the other takes `(dst, src)`.

We also think that the `(src, dst)` ordering is consistent with `std::copy()`.  You are correct that `std::copy()` takes iterators (not pointers) as parameters.  However, the order of the iterators is "src" first and then "dst".

**Pull request merged**